### PR TITLE
[exporter/awsxray] Add span links and messenger field translation to x-ray exporter.

### PR DESCRIPTION
**Pull request by atshaw43**
**Description:**
Add link span link support and support for the messaging field.

**Link to tracking Issue:**

**Testing:**
Unit Test

**Documentation:**
We are conforming to the OTel specs.
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#specifying-links
https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/trace/messaging.yaml#L110

### CODE REVIEW

File: exporter/awsxrayexporter/internal/translator/messaging.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+const MessagingPrefix = "messaging."
+
+func makeMessaging(span ptrace.Span, attributes map[string]pcommon.Value) (map[string]pcommon.Value, map[string]interface{}) {
+	var (
+		filtered  = make(map[string]pcommon.Value)
+		messaging = make(map[string]interface{})
+	)
+
+	for key, value := range attributes {
+		if strings.HasPrefix(key, MessagingPrefix) {
```
Comment (mxiamxia) [inline exporter/awsxrayexporter/internal/translator/messaging.go:31]: is it better to iterate the messaging attributes we'd like to capture than wildcard matching?

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#messaging-attributes

File: exporter/awsxrayexporter/internal/translator/messaging.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+package translator // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsxrayexporter/internal/translator"
+
+import (
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"strings"
+)
+
+const MessagingPrefix = "messaging."
```
[the remark this answers is not in the document]
Comment (atshaw43) [reply exporter/awsxrayexporter/internal/translator/messaging.go]: Fixed

**Comment by atshaw43** [comment]
Sample from message test
{"name":"signup_aggregator","id":"5f0e12026a72bb73","start_time":1679966384.08688,"service":{"version":"semver:1.1.4"},"messaging":{"operation":"process","system":"AmazonSQS"},"trace_id":"1-642240b0-5fd29989025a4cc6a9815b6f","end_time":1679966384.3018801,"fault":false,"error":false,"throttle":false,"aws":{"eks":{"cluster_name":"production","pod":"signup_aggregator-x82ufje83","container_id":null},"xray":{"auto_instrumentation":false},"account_id":"123456789"},"metadata":{"default":{"notMessaging":"myValue","otel.resource.array.key":["foo","bar"],"otel.resource.bool.key":true,"otel.resource.cloud.account.id":"123456789","otel.resource.cloud.availability_zone":"us-east-1c","otel.resource.cloud.provider":"aws","otel.resource.cloud.region":"us-east-1","otel.resource.container.image.name":"otel/signupaggregator","otel.resource.container.image.tag":"v1","otel.resource.container.name":"signup_aggregator","otel.resource.double.key":5,"otel.resource.int.key":10,"otel.resource.k8s.cluster.name":"production","otel.resource.k8s.deployment.name":"signup_aggregator","otel.resource.k8s.namespace.name":"default","otel.resource.k8s.pod.name":"signup_aggregator-x82ufje83","otel.resource.map.key":{"key1":1,"key2":"value"},"otel.resource.service.name":"signup_aggregator","otel.resource.service.version":"semver:1.1.4","otel.resource.string.key":"string"}},"parent_id":"259db1f69215f051"}

Sample from span link test
{"name":"signup_aggregator","id":"a078144849043890","start_time":1679966532.588165,"service":{"version":"semver:1.1.4"},"links":[{"trace_id":"1-64224144-5d409fb899308898ee394784","id":"e9855b3e03b0c17d"}],"trace_id":"1-64224144-0d0bf940fcb6604674a25327","end_time":1679966532.803165,"fault":false,"error":false,"throttle":false,"aws":{"eks":{"cluster_name":"production","pod":"signup_aggregator-x82ufje83","container_id":null},"xray":{"auto_instrumentation":false},"account_id":"123456789"},"metadata":{"default":{"otel.resource.array.key":["foo","bar"],"otel.resource.bool.key":true,"otel.resource.cloud.account.id":"123456789","otel.resource.cloud.availability_zone":"us-east-1c","otel.resource.cloud.provider":"aws","otel.resource.cloud.region":"us-east-1","otel.resource.container.image.name":"otel/signupaggregator","otel.resource.container.image.tag":"v1","otel.resource.container.name":"signup_aggregator","otel.resource.double.key":5,"otel.resource.int.key":10,"otel.resource.k8s.cluster.name":"production","otel.resource.k8s.deployment.name":"signup_aggregator","otel.resource.k8s.namespace.name":"default","otel.resource.k8s.pod.name":"signup_aggregator-x82ufje83","otel.resource.map.key":{"key1":1,"key2":"value"},"otel.resource.service.name":"signup_aggregator","otel.resource.service.version":"semver:1.1.4","otel.resource.string.key":"string"}},"parent_id":"7eda46d4ebe36636"}

**Review by mxiamxia — APPROVED** [review]
LGTM

**Comment by atshaw43** [comment]
I am converting this to a draft until we are ready to merge this. We have some additional work in our backend.

**Comment by atshaw43** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

Working on some backend work.

File: exporter/awsxrayexporter/internal/translator/spanLinks.go
Comment (dmitryax) [inline exporter/awsxrayexporter/internal/translator/spanLinks.go:1]: camelCase in go files is not recommended. Please use snake_case

**Review by dmitryax — APPROVED** [review]

**Comment by mridang** [comment]
Hi @atshaw43 I've been looking into the exporter code and trying to understand where this is used. I haven't found references to any link support in the AWS Xray for Python, Node or Java and therefore I'm baffled as to where this is used.

I don't see any mention of this in the docs https://docs.aws.amazon.com/xray/latest/devguide/aws-xray-interface-api.html#xray-api-segmentdocuments.html or the associated JSON schema. 

Any help is much appreciated

**Pull request merged**